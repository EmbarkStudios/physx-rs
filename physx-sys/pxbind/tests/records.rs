struct RecordOutput {
    structgen: String,
    size_asserts: String,
    rust_decls: String,
}

fn gen_records(which: &str, to_emit: &'static [&str]) -> anyhow::Result<RecordOutput> {
    let ast = pxbind::get_parsed_ast(format!("tests/data/record/{which}"))?;

    let mut consumer = pxbind::consumer::AstConsumer::default();
    consumer.consume(&ast)?;

    let record_filter = |rb: &pxbind::consumer::RecBinding<'_>| {
        to_emit.is_empty() || to_emit.iter().any(|te| *te == rb.name)
    };

    let generator = pxbind::generator::Generator {
        record_filter: Box::new(record_filter),
        ..Default::default()
    };

    let structgen = {
        let mut sg = Vec::new();
        generator.generate_structgen(&consumer, &mut sg)?;
        String::from_utf8(sg)?
    };

    let size_asserts = {
        let mut sa = Vec::new();
        generator.generate_size_asserts(&consumer, &mut sa)?;
        String::from_utf8(sa)?
    };

    let rust_decls = {
        let mut rd = Vec::new();
        generator.generate_rust_records(&consumer, &mut rd)?;
        String::from_utf8(rd)?
    };

    Ok(RecordOutput {
        structgen,
        size_asserts,
        rust_decls,
    })
}

/// Verifies that abstract base classes with no fields are properly bound
#[test]
fn abstract_() {
    let ro = gen_records("abstract.h", &[]).unwrap();

    insta::assert_snapshot!(ro.structgen);
    insta::assert_snapshot!(ro.size_asserts);
    insta::assert_snapshot!(ro.rust_decls);
}

/// Verifies that reference fields are reflected in the size and layout, but aren't
/// actually publicly exposed
#[test]
fn ref_fields() {
    let ro = gen_records("ref_fields.h", &["PxReffyMcRefface"]).unwrap();

    insta::assert_snapshot!(ro.structgen);
    insta::assert_snapshot!(ro.size_asserts);
    insta::assert_snapshot!(ro.rust_decls);
}

/// Ensures we can generate PODs that wrap types that are only exposed publicly
/// via pointer
#[test]
fn ptr_only() {
    let ro = gen_records("ptr_only.h", &[]).unwrap();

    insta::assert_snapshot!(ro.structgen);
    insta::assert_snapshot!(ro.size_asserts);
    insta::assert_snapshot!(ro.rust_decls);
}

/// Ensures we can generate PODs for some templates types that are unfortunately
/// now part of the core API
#[test]
fn simple_templates() {
    let ro = gen_records("simple_templates.h", &[]).unwrap();

    insta::assert_snapshot!(ro.structgen);
    insta::assert_snapshot!(ro.size_asserts);
    insta::assert_snapshot!(ro.rust_decls);
}

/// Ensure we can parse templates, template specializations, and template parameters
/// and stamp out unique types for each template + template param
#[test]
fn templates() {
    let ro = gen_records(
        "templates.h",
        &[
            "PxRaycastCallback",
            "PxOverlapCallback",
            "PxSweepCallback",
            "PxRaycastBuffer",
            "PxOverlapBuffer",
        ],
    )
    .unwrap();

    insta::assert_snapshot!(ro.structgen);
    insta::assert_snapshot!(ro.size_asserts);
    insta::assert_snapshot!(ro.rust_decls);
}

/// More complicated records with multiple levels of inheritance
#[test]
fn complex() {
    let ro = gen_records("complex.h", &[]).unwrap();

    insta::assert_snapshot!(ro.structgen);
    insta::assert_snapshot!(ro.size_asserts);
    insta::assert_snapshot!(ro.rust_decls);
}
