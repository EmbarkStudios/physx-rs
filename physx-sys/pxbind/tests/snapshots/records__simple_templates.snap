---
source: physx-sys/pxbind/tests/records.rs
expression: ro.structgen
---
// Automatically generated by pxbind
#include "PxPhysicsAPI.h"

using namespace physx;

#define unsafe_offsetof(st, m) ((size_t) ( (char *)&((st *)(0))->m - (char *)0 ))
#include "structgen.hpp"

int main() {
    PodStructGen sg;
    struct physx_PxVec3_Pod: public physx::PxVec3 {
        static void dump_layout(PodStructGen& sg) {
            sg.begin_struct("physx_PxVec3_Pod", "PxVec3");
            sg.add_field("float x", "x", "f32", sizeof(float), unsafe_offsetof(physx_PxVec3_Pod, x));
            sg.add_field("float y", "y", "f32", sizeof(float), unsafe_offsetof(physx_PxVec3_Pod, y));
            sg.add_field("float z", "z", "f32", sizeof(float), unsafe_offsetof(physx_PxVec3_Pod, z));
            sg.end_struct(sizeof(physx::PxVec3));
        }
    };
    physx_PxVec3_Pod::dump_layout(sg);

    struct physx_PxVec3Padded_Pod: public physx::PxVec3Padded {
        static void dump_layout(PodStructGen& sg) {
            sg.begin_struct("physx_PxVec3Padded_Pod", "PxVec3Padded");
            sg.add_field("float x", "x", "f32", sizeof(float), unsafe_offsetof(physx_PxVec3Padded_Pod, x));
            sg.add_field("float y", "y", "f32", sizeof(float), unsafe_offsetof(physx_PxVec3Padded_Pod, y));
            sg.add_field("float z", "z", "f32", sizeof(float), unsafe_offsetof(physx_PxVec3Padded_Pod, z));
            sg.add_field("uint32_t padding", "padding", "u32", sizeof(uint32_t), unsafe_offsetof(physx_PxVec3Padded_Pod, padding));
            sg.end_struct(sizeof(physx::PxVec3Padded));
        }
    };
    physx_PxVec3Padded_Pod::dump_layout(sg);

    struct physx_PxVec4_Pod: public physx::PxVec4 {
        static void dump_layout(PodStructGen& sg) {
            sg.begin_struct("physx_PxVec4_Pod", "PxVec4");
            sg.add_field("float x", "x", "f32", sizeof(float), unsafe_offsetof(physx_PxVec4_Pod, x));
            sg.add_field("float y", "y", "f32", sizeof(float), unsafe_offsetof(physx_PxVec4_Pod, y));
            sg.add_field("float z", "z", "f32", sizeof(float), unsafe_offsetof(physx_PxVec4_Pod, z));
            sg.add_field("float w", "w", "f32", sizeof(float), unsafe_offsetof(physx_PxVec4_Pod, w));
            sg.end_struct(sizeof(physx::PxVec4));
        }
    };
    physx_PxVec4_Pod::dump_layout(sg);

    struct physx_PxMat44_Pod: public physx::PxMat44 {
        static void dump_layout(PodStructGen& sg) {
            sg.begin_struct("physx_PxMat44_Pod", "PxMat44");
            sg.add_field("physx_PxVec4_Pod column0", "column0", "glam::Vec4", sizeof(physx::PxVec4), unsafe_offsetof(physx_PxMat44_Pod, column0));
            sg.add_field("physx_PxVec4_Pod column1", "column1", "glam::Vec4", sizeof(physx::PxVec4), unsafe_offsetof(physx_PxMat44_Pod, column1));
            sg.add_field("physx_PxVec4_Pod column2", "column2", "glam::Vec4", sizeof(physx::PxVec4), unsafe_offsetof(physx_PxMat44_Pod, column2));
            sg.add_field("physx_PxVec4_Pod column3", "column3", "glam::Vec4", sizeof(physx::PxVec4), unsafe_offsetof(physx_PxMat44_Pod, column3));
            sg.end_struct(sizeof(physx::PxMat44));
        }
    };
    physx_PxMat44_Pod::dump_layout(sg);

    struct physx_PxQuat_Pod: public physx::PxQuat {
        static void dump_layout(PodStructGen& sg) {
            sg.begin_struct("physx_PxQuat_Pod", "PxQuat");
            sg.add_field("float x", "x", "f32", sizeof(float), unsafe_offsetof(physx_PxQuat_Pod, x));
            sg.add_field("float y", "y", "f32", sizeof(float), unsafe_offsetof(physx_PxQuat_Pod, y));
            sg.add_field("float z", "z", "f32", sizeof(float), unsafe_offsetof(physx_PxQuat_Pod, z));
            sg.add_field("float w", "w", "f32", sizeof(float), unsafe_offsetof(physx_PxQuat_Pod, w));
            sg.end_struct(sizeof(physx::PxQuat));
        }
    };
    physx_PxQuat_Pod::dump_layout(sg);

    sg.finish();
}

